src/FuseGPUThreadLoops.cpp

```diff
@@ -1093,6 +1049,155 @@ class ExtractRegisterAllocations : public IRMutator {
+                             op->for_type, op->device_api, body);
+        } else {
+            return IRMutator::visit(op);
+        }
+    }
+
+    Stmt visit(const Store *op) override {
+        debug(4) << "Encountered store to " << op->name << "\n";
+        auto mem_type = memory_type_for_name(op->name);
+        switch (mem_type) {
+        case MemoryType::GPUShared:
+            debug(4) << "   memory type is shared\n";
+            shared_stores.insert(op->name);
+            break;
+        case MemoryType::Auto:
+        case MemoryType::Heap:
+            debug(4) << "   memory type is heap or auto\n";
+            device_stores.insert(op->name);
+            break;
+        case MemoryType::Stack:
+        case MemoryType::Register:
+        case MemoryType::LockedCache:
+        case MemoryType::VTCM:
+            break;
+        }
+
+        return IRMutator::visit(op);
+    }
+
+    Expr visit(const Load *op) override {
+        debug(4) << "Encountered load from " << op->name << "\n";
+        auto mem_type = memory_type_for_name(op->name);
+        switch (mem_type) {
+        case MemoryType::GPUShared:
+            debug(4) << "   memory type is shared\n";
+            shared_loads.insert(op->name);
+            break;
+        case MemoryType::Auto:
+        case MemoryType::Heap:
+            debug(4) << "   memory type is heap or auto\n";
+            device_loads.insert(op->name);
+            break;
+        case MemoryType::Stack:
+        case MemoryType::Register:
+        case MemoryType::LockedCache:
+        case MemoryType::VTCM:
+            break;
+        }
+
+        return IRMutator::visit(op);
+        return op;
+    }
+
+    Stmt visit(const Block *op) override {
+        if (!in_threads && op->rest.defined()) {
+            // First, we record which loads from shared/device memory occur
+            // in the rest block
+            Stmt rest = mutate(op->rest);
+
+            // Now, record which stores occur in the first stmt
+            // of this block
+            shared_stores.clear();
+            device_stores.clear();
+            Stmt first = mutate(op->first);
+
+            // If there are any loads in the rest part that
+            // load from something stored in first, insert the appropriate
+            // fence type
+            int mask = 0;
+            for (auto &st : shared_stores) {
+                auto elem = shared_loads.find(st);
+                if (elem != shared_loads.end()) {
+                    mask |= CodeGen_GPU_Dev::MemoryFenceType::Shared;
+                    break;
+                }
+            }
+            for (auto &st : device_stores) {
+                auto elem = device_loads.find(st);
+                if (elem != device_loads.end()) {
+                    mask |= CodeGen_GPU_Dev::MemoryFenceType::Device;
+                    break;
+                }
+            }
+            return Block::make(Block::make(first, make_barrier(mask)), rest);
```

Comment:
If "first" is a serial for loop and mask is zero, then it already has a barrier after each iteration, and we don't also need a barrier here. But that for loop may have things like LetStmts or ProducerConsumer nodes wrapped around it...

Is there some obvious way to avoid an extra barrier? Does it even matter?

Reply:
Not an obvious way I could think of.  Metal and PTX seem to elide extra barriers, but I'm not so confident about the other GPU languages... it may not matter.